qiskit/transpiler/passes/a_star_cx.py

```diff
@@ -0,0 +1,178 @@
+            as possible.
+        """
+        from timeit import default_timer as timer
+        from qiskit.transpiler.passes.group_gates import group_gates
+        from qiskit.transpiler.passes.a_star_mapper import a_star_mapper
+        from qiskit.transpiler.passes.post_mapping_optimization import optimize_gate_groups
+        from qiskit.dagcircuit import DAGCircuit
+        from qiskit._quantumregister import QuantumRegister
+        from qiskit._classicalregister import ClassicalRegister
+
+        # import time
+
+        import copy
+        from qiskit.mapper import Coupling, coupling_list2dict
+        from qiskit import qasm, unroll
+        # import networkx as nx
+
+        if self.gate_costs is None:
+            self.gate_costs = {
+                "id": 0,
+                "u1": 0,
+                "measure": 0,
+                "reset": 0,
+                "barrier": 0,
+                "u2": 1,
+                "u3": 1,
+                "U": 1,
+                "cx": 10,
+                "CX": 10,
+            }
+        compiled_dag = DAGCircuit()
+        compiled_dag = copy.deepcopy(dag)
+
+        # temporary circuit to add all used gates to the available gate set
+        tmp_qasm = (
+            "OPENQASM 2.0;\n"
+            + "gate cx c,t { CX c,t; }\n"
+            + "gate u3(theta,phi,lambda) q { U(theta,phi,lambda) q; }\n"
+            + "gate u2(phi,lambda) q { U(pi/2,phi,lambda) q; }\n"
+            + "gate u1(lambda) q { U(0,0,lambda) q; }\n"
+            + "qreg q[2];\n"
+            + "cx q[0], q[1];\n"
+            + "u3(0.1,0.4,0.7) q[0];\n"
+            + "u2(0.1,0.4) q[0]\n;"
+            + "u1(0.1) q[0];\n"
+        )
+        unrolled = unroll.Unroller(
+            qasm.Qasm(data=tmp_qasm).parse(),
+            unroll.DAGBackend(["cx", "u3", "u2", "u1"]),
+        )
+        tmp_circuit = unrolled.execute()
+
+        # prepare empty circuit for the result
+        empty_dag = DAGCircuit()
+        coupling = Coupling(coupling_list2dict(self.coupling_map))
+        # Note: Works with a single register named 'q' (not with 'q0' for example)
+        empty_dag.add_qreg(QuantumRegister(coupling.size(), "q"))
+        for k, v in sorted(compiled_dag.cregs.items()):
+            empty_dag.add_creg(ClassicalRegister(v, k))
+        start = timer()
+        empty_dag.basis = compiled_dag._make_union_basis(tmp_circuit)
+        empty_dag.gates = compiled_dag._make_union_gates(tmp_circuit)
+        grouped_gates = group_gates(compiled_dag)
+        # call mapper (based on an A* search) to satisfy the constraints for CNOTs
+        # given by the coupling_map
+        compiled_dag = a_star_mapper.a_star_mapper(
+            copy.deepcopy(grouped_gates),
+            coupling_list2dict(self.coupling_map),
+            coupling.size(),
+            copy.deepcopy(empty_dag),
+        )
+        grouped_gates_compiled = group_gates(compiled_dag)
+
+        # estimate the cost of the mapped circuit:
```

Comment:
Why estimated cost and not just run with a timeout? This could be a good opportunity to use `threading.Timer`